database/src/ops/block.rs

```diff
@@ -16,18 +25,149 @@ use crate::{
+        env.open_db_rw::<BlockInfoV2s>(tx_rw)?.put(
+            &block.height,
+            &BlockInfoV2 {
+                timestamp: block.block.header.timestamp,
+                total_generated_coins: block.generated_coins,
+                weight: block.weight as u64, // TODO
+                cumulative_difficulty: block.cumulative_difficulty as u64, // TODO
+                block_hash: block.block_hash,
+                cumulative_rct_outs: todo!(), // TODO
+            },
+        )
+    } else {
+        env.open_db_rw::<BlockInfoV3s>(tx_rw)?.put(
+            &block.height,
+            &BlockInfoV3 {
+                timestamp: block.block.header.timestamp,
+                total_generated_coins: block.generated_coins,
+                weight: block.weight as u64, // TODO
+                cumulative_difficulty: block.cumulative_difficulty,
+                block_hash: block.block_hash,
+                cumulative_rct_outs: todo!(),                    // TODO
+                long_term_weight: block.long_term_weight as u64, // TODO
+            },
+        )
+    }?;
+
+    // BlockBlobs: BlockHeight => BlockBlob
+    // TODO: what is a block blob in Cuprate's case?
+    env.open_db_rw::<BlockBlobs>(tx_rw)?
+        .put(&block.height, todo!())?;
+
+    // BlockHeights: BlockHash => BlockHeight
+    env.open_db_rw::<BlockHeights>(tx_rw)?
+        .put(&block.block_hash, &block.height)?;
+
+    // KeyImages: KeyImage > ()
+    {
+        let mut table = env.open_db_rw::<KeyImages>(tx_rw)?;
+        let key_images: std::slice::Iter<'_, KeyImage> = todo!();
+        for key_image in key_images {
+            table.put(key_image, &())?;
+        }
+    }
+
+    // Transaction data.
+    //
+    // - NumOutputs:      Amount         => AmountIndex
+    // - PrunedTxBlobs:   TxId           => PrunableBlob
+    // - PrunableHashes:  TxId           => PrunableHash
+    // - Outputs:         PreRctOutputId => Output
+    // - PrunableTxBlobs: TxId           => PrunableBlob
+    // - RctOutputs:      AmountIndex    => RctOutput
+    // - TxIds:           TxHash         => TxId
+    {
+        // FIXME: These will have to be re-opened write
+        // table since `open_db_rw` takes `&mut TxRw`.
+        let mut table_num_outputs = env.open_db_rw::<NumOutputs>(tx_rw)?;
+        let mut table_pruned_tx_blobs = env.open_db_rw::<PrunedTxBlobs>(tx_rw)?;
+        let mut table_prunable_hashes = env.open_db_rw::<PrunableHashes>(tx_rw)?;
+        let mut table_outputs = env.open_db_rw::<Outputs>(tx_rw)?;
+        let mut table_prunable_tx_blobs = env.open_db_rw::<PrunableTxBlobs>(tx_rw)?;
+        let mut table_rct_outputs = env.open_db_rw::<RctOutputs>(tx_rw)?;
+        let mut table_tx_ids = env.open_db_rw::<TxIds>(tx_rw)?;
+        let mut table_tx_heights = env.open_db_rw::<TxHeights>(tx_rw)?;
+        let mut table_tx_unlock_time = env.open_db_rw::<TxUnlockTime>(tx_rw)?;
+
+        for tx in block.txs {
```

Comment:
That is very annoying. Wouldn't `iter()`ate over `txs` at every table smh possible ? I wonder what would be the worst performance scenario between reopening each table every transactions, or the reiterating over each transactions at every table.

Reply:
I also feel that you will need to declare sub-functions for keeping clarity. Like monerod did. There are a lot of tables to deal with...

Reply:
Yeah, this is distinctly a problem with `cuparate_database` as:
- `redb`'s table type itself implements [`put()` with `&self`](https://docs.rs/redb/latest/redb/struct.Table.html#method.insert)
- `heed`'s table type requires a [`&mut TxRw` as input](https://docs.rs/heed/0.20.0-alpha.9/heed/struct.Database.html#method.put)

and in order to abstract over both, our custom `HeedTableRw` has to hold onto a `&mut TxRw`.

> I also feel that you will need to declare sub-functions for keeping clarity

If these tables are only added to here, I think keeping it one body is more clear.

Reply:
We could make our current `trait DatabaseR{o,w}` functions take in a TX instead?

@Boog900 what do you think?

Reply:
redb & heed both allow opening a table with a &tx. The problem for us is that because redb tracks the tx in the table and heed doesn't we need to track the tx for heed and it needs to be mutable.

Knowing this we could wrap heeds rw txs in a RefCell (an UnsafeCell would also be fine but the performance difference would probably be negligible here).

So the method to open could be moved to take an &tx and heeds DB would store a `&RefCell<RwTx>` (notice not a mutable reference) and when it comes time to use the tx it gets a multiple reference.

This should be fine as long as:
- Txs are not `Sync` (they are not)
- A table operation does not hold a reference to the inner cell once the call is over, unless the return also holds a lifetime (This is also true)

@hinto-janai what do you think?

Reply:
I sent the message before seeing yours but it still stands.

> We could make our current trait DatabaseR{o,w} functions take in a TX instead?

this would remove the atomicity of adding a block right? if we make a tx per table ...

Reply:
RefCell causes temporary value issues since it creates a temporary `Ref<'_>` before getting to the `heed::RoTx`.
```rust
fn get_range<'a, Range>(
    &'a self,
    range: Range,
) -> Result<impl Iterator<Item = Result<T::Value, RuntimeError>> + 'a, RuntimeError>
where
    Range: RangeBounds<T::Key> + 'a,
{
    //                 cannot return, temporary value
    //                         created here
    //                              v
    get_range::<T, Range>(&self.db, &self.tx_rw.borrow(), range)
}
```
If underlying `get_range()` is changed, it'll error as well, it'll force our `TxRo` to be using `RefCell` as well too.

Could use [`unsafe { try_borrow_unguarded() }`](https://doc.rust-lang.org/std/cell/struct.RefCell.html#method.try_borrow) to get around this.

> this would remove the atomicity of adding a block right?

I was thinking 1 write per all table but `open_db_rw(&mut TxRw)` wouldn't allow that - and yeah we wouldn't be able to `commit()` multiple transactions atomically...

Reply:
can we inline these functions? they are all one liners.

Reply:
wait no, that wont work :(

Reply:
This works, is it safe?
```rust
fn get_range<'a, Range>(
    &'a self,
    range: Range,
) -> Result<impl Iterator<Item = Result<T::Value, RuntimeError>> + 'a, RuntimeError>
where
    Range: RangeBounds<T::Key> + 'a,
{
    // SAFETY:
    // - our lifetime is bounded to `&'a self`
    // - `&self` means there are no mutable references
    // - we're creating/dereferencing a pointer to a good reference
    Ok(self
        .db
        .range(unsafe { &*self.tx_rw.as_ptr() }, &range)?
        .map(|res| Ok(res?.1)))
}
```

Reply:
I would just use `UnsafeCell` but yes that should be safe

Reply:
I would really rather used RefCell for runtime checks though rather than unsafe if you can think of a way ....

Reply:
this seems like a value guard issue again, I can't see an easy way to use RefCell so UnsafeCell it is I guess.

I will look at integrating miri into CI if we do decide to do this (#82)

Reply:
I have realized this is not safe!

Although it would be safe if we just had open a single table if we had open 2 tables made and iterator out of one and put values into the other while the iterator is still live we would have an active mutable and immutable reference to the same thing.

I don't think we can solve this, with `RefCell` it would cause a panic so it wouldn't be too bad but we cant use `RefCell`.

We could just ban the use of iterators with RW txs, I can't see a use for RW iters in Cuprate so it wouldn't be too much of a loss, this would also allow us to use `RefCell` again.

Reply:
How do we feel about just leaving this as is?

The cons are:
1. Performance
2. Cumbersome for callers

Counter-arguments:
1. Based on these [benchmarks](https://gist.github.com/hinto-janai/546d6308643f515a95ec749445963464), even using `redb`'s `open_db_rw()`, opening+closing a table takes `0.000000264s`, or `1.319s` for `5,000,000` calls - it'll be a tiny fraction of total time spent in something like `add_block()`
2. Users of `cuprate_database` aren't mutating tables directly anyway, they are using `ops/` - so _we_, `cuprate_database` itself, are the only ones dealing with these uglier calls - which I'm fine with writing/maintaining

Reply:
It's not awful but I would still prefer to not have this restriction.

If it makes it nicer we could also change the definition of the DatabaseRw from having a requirement of DatabaseRo to instead defining the read functions possible for DatabaseRw in the trait itself.

Reply:
264ns isn't a lot on its own. If we take [this block for example](https://xmrchain.net/block/3119473), It'll be 9*264*156 =  370656 ns or 370 µs waste at opening tables. So yes this isn't a lot, but I'm genuinely suggesting opening tables once and iterating over `txs` 9 times for a total of 9*156 tx instead of opening 9*156 tables.

Reply:
https://github.com/Cuprate/cuprate/pull/104